Legobot/Connectors/Discord.py

```diff
@@ -0,0 +1,318 @@
+
+    def identify(self, token):
+        """
+        Identifies to the websocket endpoint
+
+        Args:
+            token (string): Discord bot token
+        """
+
+        payload = {
+            'op': 2,
+            'd': {
+                'token': self.token,
+                'properties': {
+                    '$os': sys.platform,
+                    '$browser': 'legobot',
+                    '$device': 'legobot'
+                },
+                'compress': False,
+                'large_threshold': 250
+            }
+        }
+        payload['d']['synced_guilds'] = []
+        logger.info("Identifying with the following message: \
+                    {}".format(payload))
+        self.ws.send(json.dumps(payload))
+        return
+
+    def on_hello(self, message):
+        """
+        Runs on a hello event from websocket connection
+
+        Args:
+            message (dict): Full message from Discord websocket connection"
+        """
+
+        logger.info("Got a hello")
+        self.identify(self.token)
+        self.heartbeat_thread = Heartbeat(self.ws,
+                                          message['d']['heartbeat_interval'])
+        self.heartbeat_thread.start()
+        return
+
+    def on_heartbeat(self, message):
+        """
+        Runs on a heartbeat event from websocket connection
+
+        Args:
+            message (dict): Full message from Discord websocket connection"
+        """
+
+        logger.info("Got a heartbeat")
+        logger.info("Heartbeat message: {}".format(message))
+        self.heartbeat_thread.update_sequence(message['d'])
+        return
+
+    def on_message(self, message):
+        """
+        Runs on a create_message event from websocket connection
+
+        Args:
+            message (dict): Full message from Discord websocket connection"
+        """
+
+        if 'content' in message['d']:
+            metadata = self._parse_metadata(message)
+            message = Message(text=message['d']['content'],
+                              metadata=metadata).__dict__
+            logger.debug(message)
+            self.baseplate.tell(message)
+
+    def _parse_metadata(self, message):
+        """
+        Sets metadata in Legobot message
+
+        Args:
+            message (dict): Full message from Discord websocket connection"
+
+        Returns:
+            Legobot.Metadata
+        """
+
+        metadata = Metadata(source=self.actor_urn).__dict__
+        if 'author' in message['d']:
+            metadata['source_user'] = message['d']['author']['username']
+        else:
+            metadata['source_user'] = None
+        if 'channel_id' in message['d']:
+            metadata['source_channel'] = message['d']['channel_id']
+        else:
+            metadata['source_channel'] = None
+        return metadata
+
+    @staticmethod
+    def get_ws_url():
+        url = requests.get(
+            'https://discordapp.com/api/v6/gateway').json()['url']
```

Comment:
Should this be hardcoded? (Probably fine, just checking).

Reply:
Yeah, hardcoded. APIs for discord have a long support cycle and aren't very backwards compatible (it appears). Hardcoding this here means that we guarantee an API version.